pkg/controllers/metrics/state/node.go

```diff
@@ -0,0 +1,238 @@
+	daemonRequestsGaugeVec = newNodeGaugeVec("total_daemon_requests", "Node total daemon requests are the resource requested by DaemonSet pods bound to nodes. Labeled by provisioner name, node name, zone, architecture, capacity type, instance type, node phase and resource type.")
+	daemonLimitsGaugeVec   = newNodeGaugeVec("total_daemon_limits", "Node total pod limits are the resources specified by DaemonSet pod limits. Labeled by provisioner name, node name, zone, architecture, capacity type, instance type, node phase and resource type.")
+	overheadGaugeVec       = newNodeGaugeVec("system_overhead", "Node system daemon overhead are the resources reserved for system overhead, the difference between the node's capacity and allocatable values are reported by the status. Labeled by provisioner name, node name, zone, architecture, capacity type, instance type, node phase and resource type.")
+	wellKnownLabels        = getWellKnownLabels()
+)
+
+func newNodeGaugeVec(name, help string) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      name,
+			Help:      help,
+		},
+		nodeLabelNames(),
+	)
+}
+
+func nodeLabelNames() []string {
+	labels := []string{
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodePhase, // NOTE: deprecated
+	}
+
+	for _, l := range wellKnownLabels {
+		labels = append(labels, l)
+	}
+
+	return labels
+}
+
+type nodeScraper struct {
+	cluster  *state.Cluster
+	labelMap map[string]map[*prometheus.GaugeVec][]prometheus.Labels
+}
+
+func newNodeCollector(cluster *state.Cluster) *nodeScraper {
+	return &nodeScraper{
+		cluster:  cluster,
+		labelMap: make(map[string]map[*prometheus.GaugeVec][]prometheus.Labels),
+	}
+}
+
+func (ns *nodeScraper) getName() string {
+	return "nodes"
+}
+
+func (ns *nodeScraper) init(ctx context.Context) {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		crmetrics.Registry.MustRegister(gauge)
+	}
+}
+
+func (ns *nodeScraper) update(ctx context.Context) {
+	nodes := make(map[string]struct{})
+	ns.cluster.ForEachNode(func(n *state.Node) bool {
+		if _, ok := ns.labelMap[n.Node.Name]; !ok {
+			logging.FromContext(ctx).Debugf("Tracking new node: %s", n.Node.Name)
+			ns.labelMap[n.Node.Name] = make(map[*prometheus.GaugeVec][]prometheus.Labels)
+		}
+		nodes[n.Node.Name] = struct{}{}
+
+		podRequests := resources.Subtract(n.PodTotalRequests, n.DaemonSetRequested)
+		podLimits := resources.Subtract(n.PodTotalLimits, n.DaemonSetLimits)
+		allocatable := n.Node.Status.Capacity
+		if len(n.Node.Status.Allocatable) > 0 {
+			allocatable = n.Node.Status.Allocatable
+		}
```

Comment:
This is odd. Can you explain this in a comment?

Reply:
To be honest, I'm not entirely sure what this does. This was how allocatable was set originally so I kept it as is as to not accidentally break anything.